### ISSUE TITLE
Please, separate Truncated into a separate type

### ISSUE THREAD

**Issue by Anton-Latukha** [haskell-nix/hnix-store#158]
A subtask of #156 to meet #142

https://github.com/haskell-nix/hnix-store/pull/93#discussion_r547528739 mentioned that truncation should be split from hashing types. The more I work on it - the more I am aware it is a right idea.

Inclusion of the truncation as general hash type in the first place disconnected the project from hashing libraries while also depending on their interface & as a result of it imposed requirement to roll own crypto (because of creation of `data HashAlgorithm` overload).

And in the end the Truncation with `xor`in the tail to the beginning is not a widespread non-standard (seems may be even questionable) practice, all current search shows it is a merely a Nix-specific decision, so it can not be recommended generally.

Moreover, type system leverages type application of data kinds to allow any truncation length type, while in reality, truncation happens only to 20 chars. (it also was discussed, but can not find a link to it).

https://github.com/haskell-nix/hnix-store/blob/0eb0e023c9d3cd70db1e40cc154b90f67642e845/hnix-store-core/src/System/Nix/Internal/StorePath.hs#L82-L83

In short - we probably do not need the `Truncated` at all & so can be free from its rigidity & type application of data kinds. it seems that `StorePathHashAlgo` just needs to be constructed in this way, it just has a specific type & builder - that is all.

At the very least to put Truncated constructor into a separate type seems highly beneficial, simplifies the code greatly, since code starts to fully align with hashing libraries & we no longer need to roll custom type system.

**Comment by layus** [comment]
I totally agree. :+1:  from me :-)

**Comment by Anton-Latukha** [comment]
It is not that easy to split them as it looks. But I am already at the end of it, hacking the tests to preserve them.

**Comment by Anton-Latukha** [comment]
Closed by #157:

* Breaking
  * [(link)](https://github.com/haskell-nix/hnix-store/pull/157/commits/2af74986de8aef1a13dbfc955886f9935ca246a3) `System.Nix.Hash`: Base encoding/decoding function for hashes changed (due to changes in type system & separation of specially truncated Nix Store hasing):
    * `encode(InBase -> DigestWith)`
    * `decode(Base -> DigestWith)`
  * [(link)](https://github.com/haskell-nix/hnix-store/pull/157/commits/2af74986de8aef1a13dbfc955886f9935ca246a3) `System.Nix.StorePath`:
    * rm `type StorePathHashAlgo = 'Truncated 20 'SHA256` in favour of `StorePathHashPart` & `mkStorePathHashPart`.

* Additional
  * [(link)](https://github.com/haskell-nix/hnix-store/pull/157/commits/2af74986de8aef1a13dbfc955886f9935ca246a3) `System.Nix.StorePath`:
    * added `newtype StorePathHashPart = StorePathHashPart ByteString`.
      * added builder `mkStorePathHashPart :: ByteString -> StorePathHashPart`
  * [(link)](https://github.com/haskell-nix/hnix-store/pull/157/commits/2af74986de8aef1a13dbfc955886f9935ca246a3) `System.Nix.Hash`:
    * Nix store (which are specially truncated) hashes are now handled separately from other hashes:
      * add `mkStorePathHash` - a function to create a content into Nix storepath-style hash:
        `mkStorePathHash :: HashAlgorithm a => ByteString -> ByteString`
        but recommend to at once use `mkStorePathHashPart`.

It looks as a mesh, since it is split into Breaking & Additional changes.

In short, truncation moved to the `Internal.Truncation`, instead of polymorphic truncation there is a `StorePathHashPart` with the builder for it `mkStorePathHashPart :: ByteString -> StorePathHashPart` that takes any content and gives out Nix store path hash for it. Due to Nix using only concrete types & parameters for store paths, the polymorphism on the hash algorithms & of the length - got hidden away inside of the implementation, if it ever would be needed it is there in the:
https://github.com/haskell-nix/hnix-store/blob/ad4642bf4b6bd91b633473af26b660f11372c953/hnix-store-core/src/System/Nix/Internal/Hash.hs#L84-L86

As we see - now there is also a possible polymorphism on the incoming data carrier type.

**Comment by Ericson2314** [comment]
Working on C++ Nix, I am not sure truncated needs to exist at all; maybe it's just needed when doing name + hash -> store path/

**Comment by Anton-Latukha** [comment]
> Working on C++ Nix, I am not sure truncated needs to exist at all; maybe it's just needed when doing name + hash -> store path.

That is what it is. It is for `StorePath` type & that is why I put truncation into the internal module, there is no outs of it in API except for `StorePath`, `StorePathHash` & `mkStorePathHash`.
